### PR TITLE
AVRO-3666: [JAVA] Separate parsing from Schema class

### PR DESCRIPTION
**Pull request by opwvhk**
## What is the purpose of the change

Allow using pluggable parser implementations, allowing multiple formats to be parsed with the same code.

## Verifying this change

This change adds test for all added code. The change in `IDLReader` is for readability and is covered by existing tests.

Question: how can I better handle the collection parameter to the `FormattedSchemaParser` interface?

## Documentation

This code introduces a new feature: the documentation (Getting Started with Java) now refers to the new code, and the Javadoc documents the rest.

### CODE REVIEW

**Review by clesaec — COMMENTED** [review]
Nice first stepto refactor Schema parsing (full would be to externalize all static parsing function from schema to a SchemaParser like class). 
Just one question with ParseContext.namespace field.

File: lang/java/avro/src/main/java/org/apache/avro/ParseContext.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    }
+    return schema;
+  }
+
+  // Visible for testing
+  String fullName(String name, String space) {
+    int lastDot = name.lastIndexOf('.');
+    if (lastDot < 0) { // short name
+      if (!notEmpty(space)) {
+        space = namespace;
```
[the remark this answers is not in the document]
Comment (opwvhk) [reply lang/java/avro/src/main/java/org/apache/avro/ParseContext.java:184]: This is indeed one of the points where this implementation can be improved. It's already on the nomination to be removed if not needed.

File: lang/java/avro/src/main/java/org/apache/avro/ParseContext.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    PRIMITIVES.put("null", Schema.Type.NULL);
+  }
+
+  private static final Set<Schema.Type> NAMED_SCHEMA_TYPES = EnumSet.of(Schema.Type.RECORD, Schema.Type.ENUM,
+      Schema.Type.FIXED);
+  private final Map<String, Schema> oldSchemas;
+  private final Map<String, Schema> newSchemas;
+  // Visible for use in JsonSchemaParser
+  final NameValidator nameValidator;
+  private final String namespace;
```
[the remark this answers is not in the document]
Comment (opwvhk) [reply lang/java/avro/src/main/java/org/apache/avro/ParseContext.java:67]: The idea behind it was indeed to handle nested namespaces. I'm not yet certain what the best option is though: should I use this as a helping hand for parsers? Or force every parser to handle its own nesting?

Given your difficulties, I think I should opt for the latter (and remove it). What do you think?

[the remark this answers is not in the document]
Comment (clesaec) [reply lang/java/avro/src/main/java/org/apache/avro/ParseContext.java:67]: Yes, i think it's possible to remove "namespace" field from ParseContext class even without changing much current code as both Schema.parseNames... and IDLReader have their own "default namespace" stack (with function call stack for the first and  `private final Deque<String> namespaces;` field for second)

[the remark this answers is not in the document]
Comment (opwvhk) [reply lang/java/avro/src/main/java/org/apache/avro/ParseContext.java]: Fair enough. I've removed it.

**Review by clesaec — APPROVED** [review]
ParseContext::fullName method seems useless (always called with "space" argument as null ?), but, LGTM globally (and enhance current code)

**Review by clesaec — APPROVED** [review]
:+1:

File: lang/java/avro/src/test/java/org/apache/avro/util/TestSchemaResolver.java [hunk 3]
```diff
@@ -13,32 +31,25 @@
...
 
   @Test
   public void testResolving() throws IOException {
-    Path testIdl = Paths.get(".", "src", "test", "idl", "cycle.avdl").toAbsolutePath();
-    IdlReader parser = new IdlReader();
-    IdlFile idlFile = parser.parse(testIdl);
-    Protocol protocol = idlFile.getProtocol();
-    System.out.println(protocol);
-    Assert.assertEquals(5, protocol.getTypes().size());
+    // Path testIdl = Paths.get(".", "src", "test", "idl",
```
Comment (Fokko) [inline lang/java/avro/src/test/java/org/apache/avro/util/TestSchemaResolver.java:46]: Should we remove this test if it isn't being used?

**Pull request merged**